src/posix/platform/ra_listener.cpp

```diff
@@ -0,0 +1,215 @@
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the router advertisement listener.
+ */
+
+#include "ra_listener.hpp"
+
+#include <arpa/inet.h>
+#include <unistd.h>
+
+#include <openthread/border_router.h>
+#include <openthread/ip6.h>
+
+#include "platform-posix.h"
+#include "common/code_utils.hpp"
+
+#if OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE
+// ff02::01
+static const otIp6Address kRaMulticastAddress = {
+    {{0xff, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01}}};
+
+namespace ot {
+namespace Posix {
+
+enum
+{
+    kICMPv6RaType     = 134,
+    kOptionPrefixType = 3,
+};
+
+OT_TOOL_PACKED_BEGIN
+struct RaPrefixInformation
+{
+    uint8_t      mPrefixLength;
+    uint8_t      _rsv0 : 6;
+    uint8_t      mAutoConfiguration : 1;
+    uint8_t      mOnLink : 1;
+    uint32_t     mValidLifetime;
+    uint32_t     mPreferredLifetime;
+    uint32_t     _rsv1;
+    otIp6Address mPrefix;
+} OT_TOOL_PACKED_END;
+
+OT_TOOL_PACKED_BEGIN
+struct RaOptionHeader
+{
+    uint8_t mType;
+    uint8_t mLength;
+};
+
+OT_TOOL_PACKED_BEGIN
+struct RaOption
+{
+    RaOptionHeader      mHeader;
+    RaPrefixInformation mPrefixInformation;
+
+} OT_TOOL_PACKED_END;
+
+OT_TOOL_PACKED_BEGIN
+struct RaHeader
+{
+    uint8_t  mType;
+    uint8_t  mCode;
+    uint16_t mChecksum;
+    uint8_t  mHopLimit;
+    uint8_t  _rsv : 6;
+    uint8_t  mOtherConfiguration : 1;
+    uint8_t  mMananged : 1;
+    uint16_t mRouterLifetime;
+    uint32_t mReachableTime;
+    uint32_t mRetransTimer;
+} OT_TOOL_PACKED_END;
+
+RaListener::RaListener(void)
+    : mRaFd(0)
+{
+}
+
+otError RaListener::Init(unsigned int aInterfaceIndex)
+{
+    otError          error = OT_ERROR_NONE;
+    struct ipv6_mreq mreq6;
+
+    mRaFd                  = SocketWithCloseExec(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6, kSocketNonBlock);
+    mreq6.ipv6mr_interface = aInterfaceIndex;
+    memcpy(&mreq6.ipv6mr_multiaddr, kRaMulticastAddress.mFields.m8, sizeof(kRaMulticastAddress.mFields.m8));
+    VerifyOrExit(setsockopt(mRaFd, IPPROTO_IPV6, IPV6_JOIN_GROUP, &mreq6, sizeof(mreq6)) == 0, error = OT_ERROR_FAILED);
+
+exit:
+    otLogInfoPlat("RaListener::Init error=%s", otThreadErrorToString(error));
+    return error;
+}
+
+void RaListener::DeInit(void)
+{
+    if (mRaFd != 0)
+    {
+        close(mRaFd);
+        mRaFd = 0;
```

Comment:
-1?

Reply:
I think `-1` is better because `0` is `stdin`.